src/main/scala/is/hail/variant/HTSGenotypeView.scala

```diff
@@ -0,0 +1,195 @@
+package is.hail.variant
+
+import is.hail.annotations.{MemoryBuffer, RegionValue, UnsafeRow, UnsafeUtils}
+import is.hail.expr._
+import is.hail.utils._
+
+object HTSGenotypeView {
+  def apply(rowSignature: TStruct): HTSGenotypeView = {
+    rowSignature.fields(2).typ.asInstanceOf[TArray].elementType match {
+      case TGenotype => new TGenotypeView(rowSignature)
+      case _: TStruct => new StructGenotypeView(rowSignature)
+      case t => fatal(s"invalid genotype representation: $t, expect TGenotype or TStruct")
+    }
+  }
+}
+
+// FIXME: This is removed, and StructGenotypeView becomes the only genotype view when TGenotype is removed
+sealed abstract class HTSGenotypeView {
+  def setRegion(mb: MemoryBuffer, offset: Int)
+
+  def setGenotype(idx: Int)
+
+  def hasGT: Boolean
+
+  def hasAD: Boolean
+
+  def hasDP: Boolean
+
+  def hasGQ: Boolean
+
+  def hasPL: Boolean
+
+  def getGT: Int
+
+  def getAD(idx: Int): Int
+
+  def getDP: Int
+
+  def getGQ: Int
+
+  def getPL(idx: Int): Int
+}
+
+private class TGenotypeView(rs: TStruct) extends HTSGenotypeView {
+  private val ta = rs.fields(2).typ.asInstanceOf[TArray]
+  private val t = TGenotype.representation
+  private val tArrayInt32 = TArray(TInt32)
+  private val tSize = ta.elementByteSize
```

Comment:
Get rid of tSize.  I might move tArrayInt32 to HTSGenotypeView object.  Rename ta, t to tgs and tg.